### ISSUE TITLE
Reduce contract size

### ISSUE THREAD

**Issue by artob** [aurora-is-near/aurora-engine#7]
In the commit 599156f, the `release.wasm` contract size was 315,221 bytes. The subsequent commit b3be59b bloated that to 1,434,361 bytes. We'll need to figure out where the bloat comes from and cut it back down to size.

**Comment by joshuajbouw** [comment]
`let mut engine = Engine::new(*CHAIN_ID, predecessor_address());` ?

**Comment by birchmd** [comment]
[`twiggy` is a tool](https://github.com/rustwasm/twiggy) to help see what is contributing to the size of wasm files.

In our case it is easy to see that the issue is the data segment has exploded in size:

```
$ twiggy top -n 5 aurora_engine_599156f.wasm
 Shallow Bytes │ Shallow % │ Item
───────────────┼───────────┼─────────────────────
         22677 ┊     7.19% ┊ data[0]
         17065 ┊     5.41% ┊ code[88]
         12002 ┊     3.81% ┊ code[729]
          7036 ┊     2.23% ┊ code[57]
          5204 ┊     1.65% ┊ code[1036]
        251285 ┊    79.70% ┊ ... and 1289 more.
        315269 ┊   100.00% ┊ Σ [1294 Total Rows]

$ twiggy top -n 5 aurora_engine_master.wasm
 Shallow Bytes │ Shallow % │ Item
───────────────┼───────────┼─────────────────────
       1083689 ┊    71.26% ┊ data[0]
         16312 ┊     1.07% ┊ code[100]
          9316 ┊     0.61% ┊ code[637]
          7836 ┊     0.52% ┊ code[1116]
          6711 ┊     0.44% ┊ code[68]
        396816 ┊    26.09% ┊ ... and 1909 more.
       1520680 ┊   100.00% ┊ Σ [1914 Total Rows]
```

As far as I can tell, this data segment is essentially a blob which contains strings that are used for things like error messages and stack traces. For example, here is some sample text from `data[0]`:

```
/home/birchmd/.cargo/registry/src/github.com-1ecc6299db9ec823/generic-array-0.14.4/src/lib.rs
/rustc/07e0e2ec268c140e607e1ac7f49f145612d0f597/library/core/src/iter/adapters/enumerate.rs
attempt to add with overflowt
/home/birchmd/.cargo/registry/src/github.com-1ecc6299db9ec823/generic-array-0.12.4/src/lib.rs
/rustc/07e0e2ec268c140e607e1ac7f49f145612d0f597/library/alloc/src/collections/btree/map.rs
/home/birchmd/.cargo/registry/src/github.com-1ecc6299db9ec823/block-buffer-0.9.0/src/lib.rs\f9\01\10
attempt to subtract with overflow/home/birchmd/.cargo/git/checkouts/sputnikvm-97e1323bb1634338/2a8a3e9/gasometer/src/lib.rs
attempt to multiply with overflow
/home/birchmd/.cargo/git/checkouts/lunarity-b4cba36fc2978307/5201d9a/lexer/src/token.rs
/home/birchmd/.cargo/registry/src/github.com-1ecc6299db9ec823/primitive-types-0.9.0/src/lib.rs
CODE_STAGEERR_NO_UPGRADE
ERR_NOT_ALLOWED:TOO_EARLY
ERR_INVALID_TX
ERR_INVALID_CHAIN_IDERR_INVALID_ECDSA_SIGNATUREERR_META_TX_PARSE9\1b\10
ERR_NOT_ALLOWED
fatal error
error
/home/birchmd/.cargo/registry/src/github.com-1ecc6299db9ec823/memory_units-0.4.0/src/lib.rs
attempt to add with overflow\87\1c\10
attempt to subtract with overflow
attempt to multiply with overflow
/home/birchmd/.cargo/registry/src/github.com-1ecc6299db9ec823/wee_alloc-0.4.5/src/lib.rs\b4\1d\10
/home/birchmd/.cargo/registry/src/github.com-1ecc6299db9ec823/wee_alloc-0.4.5/src/imp_wasm32.rs
```

Unfortunately, I do not know why the change in https://github.com/aurora-is-near/aurora-engine/commit/b3be59b2da980c71d0a1152f582e1906c518c9c0 suddenly introduced so much more data. I'll continue investigating to see if I can find the root cause.

It is worth noting that [tools like `wasm-opt`](https://github.com/WebAssembly/binaryen#tools) do not help in this case because the data in the blob are referenced by offsets (i.e. ordinary integers) in function bodies. So pruning it is very difficult/impossible, as it would require identifying all such offsets and changing them accordingly.

**Comment by birchmd** [comment]
Update: the extra data comes from something involving the `ecrecover` function. If we comment out the logic of that function then the wasm reduces in size again

```
$ git diff
diff --git a/src/precompiles.rs b/src/precompiles.rs
index 9a0ebf0..578de8d 100644
--- a/src/precompiles.rs
+++ b/src/precompiles.rs
@@ -72,7 +72,7 @@ pub(crate) fn ecverify(hash: H256, signature: &[u8], signer: Address) -> bool {
 /// See: https://etherscan.io/address/0x0000000000000000000000000000000000000001
 #[allow(dead_code)]
 pub(crate) fn ecrecover(hash: H256, signature: &[u8]) -> Result<Address, ExitError> {
-    use sha3::Digest;
+    /*use sha3::Digest;
     assert_eq!(signature.len(), 65);

     let hash = secp256k1::Message::parse_slice(hash.as_bytes()).unwrap();
@@ -89,7 +89,7 @@ pub(crate) fn ecrecover(hash: H256, signature: &[u8]) -> Result<Address, ExitErr
             let r = sha3::Keccak256::digest(&public_key.serialize()[1..]);
             return Ok(Address::from_slice(&r[12..]));
         }
-    }
+    }*/
     Err(ExitError::Other(Borrowed("invalid ECDSA signature")))
 }

$ ls -lh target/wasm32-unknown-unknown/release/aurora_engine.wasm
-rwxrwxr-x 2 birchmd birchmd 410K Apr 15 16:45 target/wasm32-unknown-unknown/release/aurora_engine.wasm
```

Note: this is a little bigger than the 308K reported for https://github.com/aurora-is-near/aurora-engine/commit/599156fc803b3a0167f5919d8d66a3c83c49fb11 because I am working off the current `master` branch rather than going back to https://github.com/aurora-is-near/aurora-engine/commit/b3be59b2da980c71d0a1152f582e1906c518c9c0 and I guess there is 100K extra bloat that has been added since then. The point though is the wasm binary shrinks by over 1M without `ecrecover`.

The reason https://github.com/aurora-is-near/aurora-engine/commit/b3be59b2da980c71d0a1152f582e1906c518c9c0 caused such a big change in the wasm size is because it was the first commit to use `ecrecover` in a public export. Prior to that it was being discarded during compilation because it was unused.

Obviously `ecrecover` is a very important function, so cannot simply ignore it. I'll keep digging into it to see if I can isolate further were the bloat is coming from.

**Comment by birchmd** [comment]
Update: I think the large amount of data comes from `libsecp256k1`. There is a [constant defined](https://github.com/paritytech/libsecp256k1/blob/d0eedf2569bea15b453d53fa53ef95302ba1d59e/src/ecmult/mod.rs#L25) (and which is [used in the `libsecp256k1::recover` function](https://github.com/paritytech/libsecp256k1/blob/d0eedf2569bea15b453d53fa53ef95302ba1d59e/src/lib.rs#L627)) that I think will take up a lot of space.

In recent changes to the repository (not yet published to crates.io), this "static context" is now an [optional feature](https://github.com/paritytech/libsecp256k1/commit/6793ff2578321e0fe1de2e531872c50094780291), so we could try to see how well it works without the static context. This should make the wasm binary small, but probably introduce a non-trivial runtime overhead. However, I cannot test this right now because recent changes to the repository have also [broken no-std support](https://github.com/paritytech/libsecp256k1/issues/68).

In summary, I'd say we're likely stuck with a 1.5M wasm binary.

**Comment by joshuajbouw** [comment]
We could switch to Parity's crypto lib instead which I'm sure has handled contract size bloat as they too are WASM.

**Comment by joshuajbouw** [comment]
Ok, so my general understanding of this issue is that this will not be solved until https://github.com/near/nearcore/pull/3954 is merged into `nearcore`.

**Comment by joshuajbouw** [comment]
This can be closed given near/nearcore#3954 is merged in.
